### PR TITLE
[YUNIKORN-2406] Ensure task status is updated for existing allocations

### PR DESCRIPTION
**Pull request by craigcondit**
### What is this PR for?
When core accepts an existing allocation, force task state to Bound to ensure that we track state properly.

### What type of PR is it?
* [x] - Bug Fix
* [ ] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-2406

### How should this be tested?
E2e testing only; this code is deeply tied to the core implementation.

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

File: pkg/cache/scheduler_callback.go [hunk 1]
```diff
@@ -60,8 +60,15 @@ func (callback *AsyncRMCallback) UpdateAllocation(response *si.AllocationRespons
 			return err
 		}
 		if app := callback.context.GetApplication(alloc.ApplicationID); app != nil {
-			ev := NewAllocateTaskEvent(app.GetApplicationID(), alloc.AllocationKey, alloc.AllocationID, alloc.NodeID)
-			dispatcher.Dispatch(ev)
+			if task := callback.context.getTask(app.GetApplicationID(), alloc.AllocationKey); task != nil {
```
Comment (chia7712) [inline pkg/cache/scheduler_callback.go:63]: Can we get `task` from `app`? for instance: `app.GetTask(alloc.AllocationKey)`

File: pkg/cache/scheduler_callback.go [hunk 2]
```diff
@@ -60,8 +60,15 @@ func (callback *AsyncRMCallback) UpdateAllocation(response *si.AllocationRespons
...
+				if utils.IsAssignedPod(task.GetTaskPod()) {
```
Comment (chia7712) [inline pkg/cache/scheduler_callback.go:64]: Should we exclude the `failed` pod?

**Review by chia7712 — COMMENTED** [review]
@craigcondit thanks for quick fix. a couple of comments below.

For another, how to write test for this bug? I guess it needs a new e2e test? If so, a follow-up jira is needed :)

File: pkg/cache/scheduler_callback.go (hunk 1 above)
Comment (craigcondit) [reply pkg/cache/scheduler_callback.go:63]: Either method works fine and performance isn't a concern here as this happens at most once per running task which was recovered during startup. My preference would be to leave it as-is.

File: pkg/cache/task.go [hunk 3]
```diff
@@ -260,6 +260,22 @@ func (task *Task) SetTaskSchedulingState(state TaskSchedulingState) {
 	task.schedulingState = state
 }
 
+func (task *Task) MarkPreviouslyAllocated(allocationID string, nodeID string) {
+	task.sm.SetState(TaskStates().Bound)
```
[the remark this answers is not in the document]
Comment (craigcondit) [reply pkg/cache/task.go:264]: The state machine does it's own internal locking.

File: pkg/cache/scheduler_callback.go (hunk 2 above)
Comment (craigcondit) [reply pkg/cache/scheduler_callback.go:64]: Failed pods are handled via rejections and won't be here.

**Review by chia7712 — APPROVED** [review]
LGTM

**Pull request closed without merging**